### PR TITLE
Tentative acceptance protocol

### LINKED ISSUE #192: Storage Provider sends tentative acceptance
When a storage provider validates a deal, it should send a tentative acceptance. We can probably use the existing network response and adjust the statuses. 

**My exact suggested flow is as follows:**
- ValidateDeal on provider, move status to StorageDealWaitingForData and send response with status to client.
- For normal deal, client initiates data transfer and moves its own status to storagedealtransferring
- It should ALSO listen for the complete on the data transfer before it moves on to the normal read of storage deal proposal accepted
- For manual deal, client does nothing. It can move right on to the read for proposal accepted (this is problematic too but let's not address here)
- On provider, listen to data transfer events of both open AND complete. Open should move our status from StorageDealWaitingForData to StorageDealTransferring. When the request completes, we move onto StorageDealVerifyData.
- Manual DT for the provder is unchanged. We can probably remove the use of StorageDealProposalAccepted right after ValidateDealProposal and the TransferData handler entirely

### PR DESCRIPTION
**Pull request by ingar**
## Summary
Switch piece data transfer to a client push.

resolves #192

### CODE REVIEW

File: shared_testutil/testutil.go [hunk 1]
```diff
@@ -94,3 +96,12 @@ func TestVoucherEquality(t *testing.T, a, b *paych.SignedVoucher) {
 	require.NoError(t, err)
 	require.True(t, bytes.Equal(aB, bB))
 }
+
+// AssertDealState asserts equality of StorageDealStatus but with better error messaging
```
Comment (hannahhoward) [inline shared_testutil/testutil.go:100]: ❤️  Yay for this

File: storagemarket/impl/client.go [hunk 2]
```diff
@@ -372,4 +369,19 @@ func (c *clientDealEnvironment) CloseStream(proposalCid cid.Cid) error {
...
+func (c *clientDealEnvironment) StartDataTransfer(ctx context.Context, to peer.ID, voucher datatransfer.Voucher, baseCid cid.Cid, selector ipld.Node) error {
+	_, err := c.c.dataTransfer.OpenPushDataChannel(ctx, to, voucher, baseCid, selector)
+	return err
+}
+
+func (c *clientDealEnvironment) OnDataTransferEvent(ctx context.Context, eventCode datatransfer.EventCode, cb providerstates.OnDataTransferEventCallback) datatransfer.Unsubscribe {
+	// Can there be overlapping Open/Complete events?
+	return c.c.dataTransfer.SubscribeToEvents(func(event datatransfer.Event, channelState datatransfer.ChannelState) {
+		if event.Code == eventCode {
+			err := cb(event)
```
Comment (hannahhoward) [inline storagemarket/impl/client.go:381]: SubscribeToEvents will get called for every transfer possible. I don't think this is a good way to setup this listener.

You could use the approach above and add a Proposal CID and check the Voucher (also check that it's a data transfer voucher

OR

You could use a single subscriber, check the voucher, and send events in from the outside -- the same way as was used with DataTransferSubscriber on the provider.

I think this is more efficient.

And then you just avoid a StateHandler entirely for StorageDealTransferring.

File: storagemarket/impl/provider.go [hunk 3]
```diff
@@ -392,6 +388,17 @@ func (p *providerDealEnvironment) Ask() storagemarket.StorageAsk {
 	return *sask.Ask
 }
 
+func (p *providerDealEnvironment) OnDataTransferEvent(ctx context.Context, eventCode datatransfer.EventCode, cb providerstates.OnDataTransferEventCallback) datatransfer.Unsubscribe {
```
Comment (hannahhoward) [inline storagemarket/impl/provider.go:391]: same problem as client. Again, I think this is better done via a single subscriber and reading the voucher to know where to route the event

Comment (hannahhoward) [reply storagemarket/impl/provider.go:391]: it generally should work with the existing DataTransferSubscriber with only a couple modifications

File: storagemarket/impl/providerstates/provider_states.go [hunk 4]
```diff
@@ -101,35 +103,42 @@ func ValidateDealProposal(ctx fsm.Context, environment ProviderDealEnvironment,
...
 
-	return ctx.Trigger(storagemarket.ProviderEventDataTransferInitiated)
+	environment.OnDataTransferEvent(ctx.Context(), datatransfer.Open, func(event datatransfer.Event) error {
+		return ctx.Trigger(storagemarket.ProviderEventDataTransferInitiated)
+	})
+
+	return nil
+}
+
+func TransferringData(ctx fsm.Context, environment ProviderDealEnvironment, deal storagemarket.MinerDeal) error {
```
Comment (hannahhoward) [inline storagemarket/impl/providerstates/provider_states.go:136]: This handler should be unneccesary if you have a general data transfer subscriber

File: storagemarket/impl/clientstates/client_states.go [hunk 5]
```diff
@@ -77,6 +86,50 @@ func ProposeDeal(ctx fsm.Context, environment ClientDealEnvironment, deal storag
...
+	)
+
+	if err != nil {
+		return ctx.Trigger(storagemarket.ProviderEventDataTransferFailed, xerrors.Errorf("failed to open push data channel: %w", err))
+	}
+
+	return ctx.Trigger(storagemarket.ClientEventDataTransferInitiated)
+}
+
+func TransferringData(ctx fsm.Context, environment ClientDealEnvironment, deal storagemarket.ClientDeal) error {
```
Comment (hannahhoward) [inline storagemarket/impl/clientstates/client_states.go:126]: This should be unneccesary with a single subscriber for the client

File: storagemarket/impl/client.go (hunk 2 above)
Comment (ingar) [reply storagemarket/impl/client.go:400]: That makes way more sense.  I was thinking of the "resuming" scenario, but this covers all the cases.

File: storagemarket/impl/providerstates/provider_states.go [hunk 6]
```diff
@@ -45,6 +45,8 @@ type ProviderDealEnvironment interface {
 	DealAcceptanceBuffer() abi.ChainEpoch
 }
 
+type OnDataTransferEventCallback func(event datatransfer.Event) error
```
Comment (hannahhoward) [inline storagemarket/impl/providerstates/provider_states.go:48]: Is this used?

File: storagemarket/impl/providerstates/provider_fsm.go [hunk 7]
```diff
@@ -24,18 +24,18 @@ var ProviderEvents = fsm.Events{
 			deal.Message = xerrors.Errorf("deal rejected: %w", err).Error()
 			return nil
 		}),
-	fsm.Event(storagemarket.ProviderEventDealAccepted).
-		From(storagemarket.StorageDealValidating).To(storagemarket.StorageDealProposalAccepted),
+	fsm.Event(storagemarket.ProviderEventDataRequested).
```
Comment (hannahhoward) [inline storagemarket/impl/providerstates/provider_fsm.go:27]: Is ProviderEventDataRequested and ProviderEventWaitingForManualData functionally equivalent at this point? Maybe we should just combine them.

I ask cause I noticed that in the provider_states.ValidateDealProposal, at the end there is an if statement dispatching two different events, but no actual different code run other than a log message. I'd just make these a single case. (waiting for the client to send us data over data transfer is essentially the same as waiting for the client to send us a hard drive)

File: storagemarket/impl/providerstates/provider_states_test.go [hunk 8]
```diff
@@ -962,10 +916,19 @@ type fakeEnvironment struct {
 	fs                      filestore.FileStore
 	pieceStore              piecestore.PieceStore
 	dealAcceptanceBuffer    abi.ChainEpoch
+	dtCallbacks             map[datatransfer.EventCode]providerstates.OnDataTransferEventCallback
 	expectedTags            map[string]struct{}
 	receivedTags            map[string]struct{}
 }
 
+func (fe *fakeEnvironment) OnDataTransferEvent(ctx context.Context, eventCode datatransfer.EventCode, cb providerstates.OnDataTransferEventCallback) datatransfer.Unsubscribe {
```
Comment (hannahhoward) [inline storagemarket/impl/providerstates/provider_states_test.go:924]: delete?

File: storagemarket/impl/clientstates/client_states_test.go [hunk 9]
```diff
@@ -356,9 +419,36 @@ func makeNode(params nodeParams) storagemarket.StorageClientNode {
...
+	fe.startDataTransferCalls = append(fe.startDataTransferCalls, dataTransferParams{
+		to:       to,
+		voucher:  voucher,
+		baseCid:  baseCid,
+		selector: selector,
+	})
+	return nil
+}
+
+func (fe *fakeEnvironment) OnDataTransferEvent(ctx context.Context, eventCode datatransfer.EventCode, cb providerstates.OnDataTransferEventCallback) datatransfer.Unsubscribe {
```
Comment (hannahhoward) [inline storagemarket/impl/clientstates/client_states_test.go:446]: delete?

File: storagemarket/impl/client.go (hunk 2 above)
Comment (ingar) [reply storagemarket/impl/client.go:400]: Oh right.

File: storagemarket/impl/providerstates/provider_fsm.go (hunk 7 above)
Comment (ingar) [reply storagemarket/impl/providerstates/provider_fsm.go:27]: 👍

File: storagemarket/impl/clientstates/client_fsm.go [hunk 10]
```diff
@@ -34,25 +34,41 @@ var ClientEvents = fsm.Events{
...
 			return nil
 		}),
+	fsm.Event(storagemarket.ClientEventUnexpectedDealState).
+		From(storagemarket.StorageDealWaitingForDataRequest).To(storagemarket.StorageDealFailing).
+		Action(func(deal *storagemarket.ClientDeal, status storagemarket.StorageDealStatus) error {
+			deal.Message = xerrors.Errorf("unexpected deal status while waiting for data request: %d", status).Error()
+			return nil
+		}),
+	fsm.Event(storagemarket.ClientEventDataTransferFailed).
+		From(storagemarket.StorageDealWaitingForDataRequest).To(storagemarket.StorageDealFailing).
```
[the remark this answers is not in the document]
Comment (ingar) [reply storagemarket/impl/clientstates/client_fsm.go:63]: I would have never seen this.

**Pull request merged**